docs/workflow.md

```diff
@@ -516,167 +486,208 @@ results/
+A mapping closure should usually have only one parameter. However, if the incoming values are tuples, the closure can specify a parameter for each tuple element for more convenient access, also known as "destructuring" or "unpacking".
+:::
 
 ### Index files
 
-A publish target can create an index file of the values that were published. An index file is a useful way to save the metadata associated with files, and is more flexible than encoding metadata in the file path. Currently only CSV files are supported.
+A publish target can create an index file of the values that were published. An index file preserves the structure of channel values, including metadata, which is simpler than encoding this information with directories and file names. The index file can be CSV (`.csv`) or JSON (`.json`).
 
 For example:
 
 ```groovy
 workflow {
-    ch_foo = Channel.of(
-        [id: 1, name: 'foo 1'],
-        [id: 2, name: 'foo 2'],
-        [id: 3, name: 'foo 3']
+    main:
+    ch_fastq = Channel.of(
+        [ [id: 1, name: 'sample 1'], '1a.fastq', '1b.fastq' ],
+        [ [id: 2, name: 'sample 2'], '2a.fastq', '2b.fastq' ],
+        [ [id: 3, name: 'sample 3'], '3a.fastq', '3b.fastq' ]
     )
 
     publish:
-    ch_foo >> 'foo'
+    ch_fastq >> 'fastq'
 }
 
 output {
-    directory 'results'
-
-    'foo' {
+    'fastq' {
         index {
             path 'index.csv'
         }
     }
 }
 ```
 
-The above example will write the following CSV file to `results/foo/index.csv`:
+The above example will write the following CSV file to `results/fastq/index.csv`:
 
 ```csv
-"id","name"
-"1","foo 1"
-"2","foo 2"
-"3","foo 3"
+"id","name","fastq_1","fastq_2"
+"1","sample 1","results/fastq/1a.fastq","results/fastq/1b.fastq"
+"2","sample 2","results/fastq/2a.fastq","results/fastq/2b.fastq"
+"3","sample 3","results/fastq/3a.fastq","results/fastq/3b.fastq"
 ```
 
-You can customize the index file by specifying options in a block, for example:
+You can customize the index file with additional directives, for example:
 
 ```groovy
 index {
     path 'index.csv'
-    header ['name', 'extra_option']
+    header ['id', 'fastq_1', 'fastq_1']
     sep '\t'
-    mapper { val -> val + [extra_option: 'bar'] }
+    mapper { meta, fq_1, fq_2 -> meta + [fastq_1: fq_1, fastq_2: fq_2] }
 }
 ```
 
-The following options are available:
+This example will produce the same index file as above, but with the `name` column removed and with tabs instead of commas.
+
+See [Reference](#reference) for the list of available index directives.
+
+### Migrating from first preview
+
+The first preview of workflow publishing was introduced in 24.04. The second preview, introduced in 24.10, made the following breaking changes:
 
-`header`
-: When `true`, the keys of the first record are used as the column names (default: `false`). Can also be a list of column names.
+- The process `publish:` section has been removed. Channels should be published only in workflows, ideally the entry workflow.
```

Comment:
It's too confusing. If a process publishes some outputs by default and you don't want to publish them, then you have to explicitly disable them. And one of the goals of workflow outputs was to have all published outputs in one place, instead of being scattered across the pipeline

My initial feeling was to only allow publishing in the entry workflow, but that would also create a lot of boilerplate to pass all the channels up from the subworkflows. So a happy middle ground for now is to allow publishing from any workflow, but not processes

Reply:
Also, the original motivation for process publishing is to provide some sensible publishing defaults for a module. I think this is better handled by providing an entry workflow with the module:
```groovy
process ASPERA_CLI {
    input:
    // ...

    output:
    tuple val(meta), path("*fastq.gz"), emit: fastq
    tuple val(meta), path("*md5")     , emit: md5

    script:
    // ...
}

workflow {
    main:
    ASPERA_CLI ( params.input, params.args )

    publish:
    ASPERA_CLI.out.fastq >> 'fastq'
    ASPERA_CLI.out.md5 >> 'md5'
}
```

This gives an example of how to use the module, both params and publishing, but they are opt-in. The user can incorporate these publishing rules into their pipeline explicitly, if they want to.